rfcs/20221006-frontend_event_tracking.md

```diff
@@ -0,0 +1,469 @@
+    width: window.innerWidth, // The width of the user's device
+    height: window.innerHeight, // The height of the user's device
+    ipAddress: "", // Should be read from the X-Forwarded-For header server-side, will need to be stored securely during the user's session
+    userAgent: navigator.userAgent,
+  };
+  /**
+   * Actually send the event here, with some additional values.
+   * I would have prefered to use `navigator.sendBeacon` here
+   * but it is blocked by most ad blockers.
+   *
+   * We might not want to call our api route "analytics" to
+   * further avoid blocking by adblockers, but I'll use that
+   * name for clarity here.
+   */
+  const response = await fetch(`/analytics`, {
+    method: "POST",
+    headers: { "Content-Type": "application/json" },
+    body: JSON.stringify({ ...defaults, ...payload }),
+  });
+}
+```
+
+### Implementation Plan
+
+- [ ] Create "Event Tracking" GitHub milestone in the frontend repository
+- [ ] Design: design a privacy and analytics opt-out checkbox on the Privacy page. Also a pop-up banner to be shown to users letting them know Openverse uses analytics (can use the existing audio notice for this)
+- [ ] Create the `ANALYTICS_ENABLED` feature flag which defaults to `off`/`false`.
+- [ ] Create a PR which adds an `allowAnalytics` value to our UI cookie store; implements the checkbox on the privacy page; and shows a 'Openverse collects analytics information...' banner to first-time (new session) users.
+- [ ] Create an issue for the creation of the `sendEvent` function and the api route to post analytics events to. The api route will post events to an `ANALYTICS_ENDPOINT` environment variable. If that variable is unset and/or the `ANALYTICS_ENABLED` flag is off, the analytics events will not send.
+- [ ] Create each event in the sample implementation. Each event will need a detailed description of the payload and trigger (the actual user interaction that fires off the event). Events will need to account for accessibility concerns. Will keyboard users' actions trigger the events the same way as mouse users?
+- [ ] Future: Create a PR to turn `ANALYTICS_ENABLED` to `on`/`true`, set the `ANALYTICS_ENDPOINT` environment variable, and do any munging of data necessary for the decided on analytics service.
+
+## Concerns / Pitfalls
+
+- Related to content safety, this is an area where we are surfacing user information in the form of text, or even malicious URLs, to Openverse developers. These searches could include explicit or sensitive material. We're also potentially surfacing explicit image results to anyone working with this data.
+- We're building something here that is commonly drop-in when 3rd party tools are used. At least for page views; custom events still require custom work. Is this a poor use of time?
+- I think it's great to build this decoupled from a specific back-end, but if we never actually implement a backend this would largely be a wasted effort.
```

Comment:
Oh, wow. Yeah, in general, I think it's very difficult to evaluate the success of a data collection effort without looking at the data itself. Am I right in understanding that this is a data collection design that is separate from actually storing any data?

Reply:
I've commented [here](https://github.com/WordPress/openverse/pull/312#discussion_r1062735808) with some context about this question, thanks for asking!